besselk: implement series expansion for non-integer order

Kishore96in opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

<!-- #### References to other Issues or PRs -->
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Series expansion for besselk (usually denoted as $K_\nu$) about the origin.
(<https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/01/0003/>). Earlier, it was implemented only for integer orders.

This PR also fixes `besselk._eval_as_leading_term` for negative non-integer orders.

#### Other comments
I have also fixed a bug in the series expansion for integer orders (it was sometimes returning the wrong `Order` object), and added tests.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
  * Implemented besselk series expansion for non-integer order
  * besselk: fixed calculation of leading term for non-integer negative orders
<!-- END RELEASE NOTES -->

Kishore96in committed:
besselk: add series expansion for non-integer order and fix integer
order

NotImplementedError is now raised for unhandled cases (i.e. when the
order is not real.

We now also also return the correct Order when the order (n) of the
argument of the Bessel function is higher than the order requested by
the user.

For integer order:
* fix various edge cases and order
of returned term.
* Also remove unused branch of if-else: (k-nu)*k can never be zero for k in range(1,nu)

Re. the order of the returned term:
Since the O() term in a is multiplied by log(x), it is not the same as
the ones which are present (but removed in previous lines) in b and c.
This can sometimes lead to the wrong order of the truncated terms being
indicated in the final expression.

Kishore96in committed:
besselk: add tests: series expansion for integer order

These earlier did not work because the wrong Order object was being
returned

Also added an edge case that tests if the series truncation is correct.

anutosh491 commented on sympy/functions/special/bessel.py:
```diff
@@ -690,51 +690,70 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
+                b, c = [], []
+                o = Order(x**n, x)
+                r = (z/2)._eval_nseries(x, n, logx, cdir).removeO()
+                if r is S.Zero:
```
The overall impelemtation looks correct to me. Can you let me know of a series expansion example where this change comes in though ?

Kishore96in replied:
On master:
```
In [3]: besselk(2,x**2).series(x, n=2)
Out[3]: 
 ⎛ 2⎞
O⎝x ⎠
```

With this fix:
```
In [1]: besselk(2,x**2).series(x, n=2)
Out[1]: 
2     ⎛ 2⎞
── + O⎝x ⎠
 4        
x   
```

First sympy gets `x**2 -> O(x**2)`, but on putting this into the expansion of the Bessel function, we get `besselk(2,x**2) -> O(x**(-4))`. `Expr.series` notices that we did not get the requested order, and so increases the order of the intermediate series evaluations to get the answer shown above.

This also fixes the case `(x**2*besselk(2, x)).series(x ,n=2) == 2 + O(x**2)` (which I have added as a test).
Should I add the simpler case above to the tests as well?

anutosh491 replied:
Correct, I do get your point now . We are missing out on some terms . I would say add the minimalistic example first and then you could add a couple more tests to top that. 
I realize we would find also similar problems in `bessely` too as I have pointed here (https://github.com/sympy/sympy/pull/25664#issuecomment-1713248527) . Would be interested in fixing that too ? Once we get this PR merged, that should be easy for you .

Kishore96in replied:
Added the test.

anutosh491 commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -166,7 +166,13 @@ def test_besselk_series():
     assert besselk(-2, sin(x)).series(x, n=4) == besselk(2, sin(x)).series(x, n=4)
+    assert (x**2*besselk(2, x)).series(x ,n=2) == 2 + O(x**2)
 
+def test_besselk_frac_order_series():
```
Also could you please refactor the way, you've framed the test . Look at `test_besselk_series` for how to frame tests cleanly.

Kishore96in replied:
I've added newlines to the long expressions. Is there something else you had in mind?

anutosh491 commented:
Thanks @Kishore96in, good catch. I think so the non integer cases were being ignored . I looked at the tests you've added and they looked just fine to me. cc @jksuom could you also confirm ?

anutosh491 commented:
Now that you've caught this, would you be interested to implement non-integer order series expansions for the `bessely` function ? If yes, go ahead and do it or else open an issue for it so that we can keep track of it. Use https://functions.wolfram.com/Bessel-TypeFunctions/BesselY/06/01/04/01/01/0003/ for reference

anutosh491 commented:
The `_eval_as_leading_term` function works correctly according to me for these cases too, but I would like if you could check that for me.

anutosh491 commented on sympy/functions/special/bessel.py:
```diff
@@ -693,66 +693,67 @@ def _eval_nseries(self, x, n, logx, cdir=0):
+                return a + Add(*b) + Add(*c) # Order term comes from a
+            else:
+                #Reference: https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/01/0003/ (only for non-integer order)
+                r = (z/2)._eval_nseries(x, n, logx, cdir).removeO()
```
I see the variable `r` being computed is similar for both if-else branches. This duplication can be taken care of.

Kishore96in replied:
Done. I have also added a call to `_mexpand` on each term of the series (just like in the integer case).

Kishore96in commented:
> Now that you've caught this, would you be interested to implement non-integer order series expansions for the `bessely` function ? If yes, go ahead and do it or else open an issue for it so that we can keep track of it. Use https://functions.wolfram.com/Bessel-TypeFunctions/BesselY/06/01/04/01/01/0003/ for reference

Same PR or a different one?

anutosh491 commented:
Try it through the next PR . I would like you to address all the changes you did here for `bessely` too. Once you've addressed the changes I've pointed above, I would also like you to squash the commits and have one or two concrete commits.

Kishore96in commented:
> The `_eval_as_leading_term` function works correctly according to me for these cases too, but I would like if you could check that for me.

There seems to be a problem with negative non-integer orders.
```
In [26]: besselk(-S(5)/3,x).as_leading_term(x)
Out[26]: 
     3 ____  2/3                    3 ____  2/3              3 ____  2/3  
- 12⋅╲╱ -1 ⋅2   ⋅log(x)⋅(-5/3)! + 9⋅╲╱ -1 ⋅2   ⋅Γ(-2/3) + 12⋅╲╱ -1 ⋅2   ⋅l
──────────────────────────────────────────────────────────────────────────
                                                                          
                                                                          

                3 ____  2/3                   3 ____  2/3               3 
og(2)⋅(-5/3)! - ╲╱ -1 ⋅2   ⋅√3⋅π⋅Γ(-2/3) - 12⋅╲╱ -1 ⋅2   ⋅γ⋅Γ(-2/3) - 9⋅╲╱
──────────────────────────────────────────────────────────────────────────
    5/3                                                                   
 6⋅x   ⋅(-5/3)!⋅Γ(-2/3)                                                   

____  2/3               
 -1 ⋅2   ⋅log(3)⋅Γ(-2/3)
────────────────────────
                        
                        

In [27]: besselk(S(5)/3,x).as_leading_term(x)
Out[27]: 
 2/3       
2   ⋅(2/3)!
───────────
     5/3   
    x      
```

I would expect both the above to return the same answer (since $K_\nu(z) = K_{-\nu}(z)$; equation 9.6.6 of Abramowitz-Stegun).

Kishore96in committed:
besselk._eval_as_leading_term: handle additional cases

* where the arg approaches zero but the order is negative non-integer
* where the arg approaches a constant
* arg with negative exponent (expansion about infinity)

Also took the opportunity to simplify the code a bit. An error is raised
if we can't figure out the zero-ness of the order

Kishore96in committed:
besselk: leading_term: add tests for recently fixed cases

Also compare with series computation

Kishore96in commented:
@anutosh491 I've now added a couple of commits fixing `besselk._eval_as_leading_term` for negative non-integer orders, and also added some tests to `test_besselk_leading_term`.

Kishore96in commented on sympy/functions/special/bessel.py:
```diff
@@ -677,64 +677,79 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
-            return arg
+            if nu is S.Zero:
+                #Equation 9.6.8 of Abramowitz and Stegun (10th ed, 1972).
+                term = -log(z) - S.EulerGamma + log(2)
```
@anutosh491 The reference I've quoted actually gives this without the constant term, but I've added the constant to avoid breaking an earlier test. It is unclear to me why Sympy says
```
In [15]: (1+log(x)).as_leading_term(x)
Out[15]: log(x) + 1
```

anutosh491 replied:
Yes we do maintain the constants while calculating the leading terms for log based functions. Something like 
`log(2*x).as_leading_term(x, cdir=1)` can be further split into `(log(x) + log(2)).as_leading_term(x, cdir=1)` if that makes sense and we do need to take the log(2) term into account here for various reasons. Check out the examples/issues fixed here (#23592) if that further interests you but as of now I think what you did is correct, maintain the constant too.

anutosh491 commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -69,6 +69,8 @@ def test_besselk_leading_term():
     assert besselk(0, x).as_leading_term(x) == -log(x) - S.EulerGamma + log(2)
     assert besselk(1, sin(x)).as_leading_term(x) == 1/x
     assert besselk(1, 2*sqrt(x)).as_leading_term(x) == 1/(2*sqrt(x))
+    assert besselk(S(5)/3,x).as_leading_term(x) == 2**(S(2)/3)*gamma(S(5)/3)/x**(S(5)/3)
```
Give single space after every `,` you use to maintain PEP 8 style of python code

anutosh491 commented:
Also update the release notes with whatever you are fixing here. Now that you've fixed some cases for leading term too, add that too. Also have you checked series expansion for these cases (**negative non-integer orders**). Do that and get back. I would say also add an example or two as a test.

anutosh491 commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -166,7 +168,20 @@ def test_besselk_series():
         S.EulerGamma/2) + x**(S(5)/2)*(log(x)/24 - S(5)/36 + S.EulerGamma/12) + O(x**3*log(x))
     assert besselk(-2, sin(x)).series(x, n=4) == besselk(2, sin(x)).series(x, n=4)
-
+    assert besselk(2,x**2).series(x, n=2) == 2/x**4 + O(x**2)
```
Space after `,`

Kishore96in commented:
> Also update the release notes with whatever you are fixing here. Now that you've fixed some cases for leading term too, add that too. 

Done.

> Also have you checked series expansion for these cases (**negative non-integer orders**). Do that and get back. I would say also add an example or two as a test. 

The series expansion we now use for non-integer orders manifestly exhibits the symmetry $K_\nu(z) = K_{-\nu}(z)$, but I've now added a couple of tests for negative non-integer orders.

> I think so all these things would also be need to be fixed for the `bessely` function, so go ahead and open an issue for it. It is always good to have an issue that we can track and which gives us an overview of whatever is going wrong. We don't really have an issue for `besselk` here but that's okay for now.

#25678

Kishore96in commented:
I've just squashed and rebased the commits in this PR.

anutosh491 commented on sympy/functions/special/bessel.py:
```diff
@@ -677,64 +677,79 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
-            arg = Add(*[term_one, term_two, term_three]).as_leading_term(x, logx=logx)
-            return arg
+            if nu is S.Zero:
+                #Equation 9.6.8 of Abramowitz and Stegun (10th ed, 1972).
```
Give a space after `#` everytime you put a reference

anutosh491 commented on sympy/functions/special/bessel.py:
```diff
@@ -677,64 +677,79 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
-                    b.append(term)
+            o = Order(x**n, x)
+            if nu.is_integer:
+                #Reference: https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/ (only for integer order)
```
Here too give a space

anutosh491 commented on sympy/functions/special/bessel.py:
```diff
@@ -677,64 +677,79 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
+                term = -log(z) - S.EulerGamma + log(2)
+            else:
+                #Equation 9.6.9 of Abramowitz and Stegun (10th ed, 1972).
+                term = gamma(Abs(nu))*(z/2)**(-Abs(nu))/2
```
I believe the series (`_eval_nseries`) implementation is correct. But I am a bit skeptical about this. Though the reference seems correct to me but can we have a series implementation where we extract the leading term out of all of them. We currently use https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/ ( when `nu` is a natural number) but I think we need we can refer to this link (https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/01/) when `nu` is not integer. 
Could you have a look into this . 
We could have two series based implementations, one when `nu` is integer and one when `nu` is not equal to integer

Kishore96in replied:
If we calculate the leading term for the two separate series expansions for integer and non-integer `nu` (basically just plug $k=0$ in both <https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/01/0003/> and <https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/> and then choose the $z^{-\nu}$ term when $\nu\ne 0$), it does come out to be what I have currently implemented.

I could do the leading term calculation in Sympy (as was done earlier) instead of doing it manually and putting the result, but that just looks like it would complicate the code (introducing another pair of if-branches that end up with mathematically identical results) without improving readability or speed.

If you insist, I can make the change you suggest, but I just wanted to confirm before proceeding.

anutosh491 replied:
Hmm, I would think on this and get back to you. I do think your approach would work equally fine but I would like to make sure we don't miss out on any cases which are symbolic in nature. Many times when we simplify things, we end up loosing on the symbolic aspect on things for eg we should also take care of any general `besselk(nu, z)` case where no numeric stuff is involved and it's all symbolic . Hence a substitution might look obvious to us numerically but we might miss out on things !
Actually why don't you go ahead and test out your code on some symbolic cases and series expansions.

Kishore96in replied:
> Actually why don't you go ahead and test out your code on some symbolic cases and series expansions.

Do you mean something like the following?
```
assert besselk(n, x).as_leading_term(x).subs({n:S(5)/7}) == besselk(S(5)/7, x).series(x).as_leading_term(x)
assert besselk(n, x).as_leading_term(x).subs({n:3}) == besselk(3, x).series(x).as_leading_term(x)
```

These currently work, and I can add them as tests if required.

anutosh491 replied:
I think so, yeah something like this should work (also for series I reckon) but if you say that series expansions don't work when the order is symbolic in nature we could keep it like that for now and take that up in some other pr

Kishore96in replied:
I just checked the symbolic case. On master, we get wrong results:
```
In [6]: besselk(y, x).as_leading_term(x).subs({y:2})
Out[6]: 
   2             2    2              2       
  x ⋅log(x)   γ⋅x    x ⋅(3/2 - γ)   x ⋅log(2)
- ───────── - ──── + ──────────── + ─────────
      8        16         16            8    

In [7]: besselk(2, x).as_leading_term(x)
Out[7]: 
2 
──
 2
x 
```
(both the results above should've been the same; the second one is the correct answer)
With this PR, we get
```
In [1]: besselk(y, x).as_leading_term(x).subs({y:2})
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[1], line 1
----> 1 besselk(y, x).as_leading_term(x).subs({y:2})

File /usr/lib/python3.12/site-packages/sympy/core/cache.py:72, in __cacheit.<locals>.func_wrapper.<locals>.wrapper(*args, **kwargs)
     69 @wraps(func)
     70 def wrapper(*args, **kwargs):
     71     try:
---> 72         retval = cfunc(*args, **kwargs)
     73     except TypeError as e:
     74         if not e.args or not e.args[0].startswith('unhashable type:'):

File /usr/lib/python3.12/site-packages/sympy/core/expr.py:3496, in Expr.as_leading_term(self, logx, cdir, *symbols)
   3494 if x not in self.free_symbols:
   3495     return self
-> 3496 obj = self._eval_as_leading_term(x, logx=logx, cdir=cdir)
   3497 if obj is not None:
   3498     from sympy.simplify.powsimp import powsimp

File /usr/lib/python3.12/site-packages/sympy/functions/special/bessel.py:708, in besselk._eval_as_leading_term(self, x, logx, cdir)
    706         term = gamma(Abs(nu))*(z/2)**(-Abs(nu))/2
    707     else:
--> 708         raise NotImplementedError(f"Cannot proceed without knowing if {nu} is zero or not.")
    710     return term.as_leading_term(x, logx=logx)
    711 elif e.is_negative:
    712     # Equation 9.7.2 of Abramowitz and Stegun (10th ed, 1972).

NotImplementedError: Cannot proceed without knowing if y is zero or not.
```

anutosh491 replied:
I think that might do.

Kishore96in commented:
There is still an issue with series expansions when the order is symbolic. The same issue is there with `besselj`, but is it something that should be fixed? Ideally, I would expect some sort of Piecewise object (depending on the order of the bessel function) to be returned, but I suppose that can wait for another PR. Shall I open an issue to track that?

```
In [6]: besselk(n,x).series(x)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File /usr/lib/python3.11/site-packages/sympy/core/mul.py:1968, in Mul._eval_nseries(self, x, n, logx, cdir)
   1967     else:
-> 1968         raise ValueError
   1970 n0 = sum(t[1] for t in ords if t[1].is_number)

ValueError: 

During handling of the above exception, another exception occurred:

NotImplementedError                       Traceback (most recent call last)
Cell In[6], line 1
----> 1 besselk(n,x).series(x)

File /usr/lib/python3.11/site-packages/sympy/core/expr.py:3026, in Expr.series(self, x, x0, n, dir, logx, cdir)
   3023 if x.is_positive is x.is_negative is None or x.is_Symbol is not True:
   3024     # replace x with an x that has a positive assumption
   3025     xpos = Dummy('x', positive=True)
-> 3026     rv = self.subs(x, xpos).series(xpos, x0, n, dir, logx=logx, cdir=cdir)
   3027     if n is None:
   3028         return (s.subs(xpos, x) for s in rv)

File /usr/lib/python3.11/site-packages/sympy/core/expr.py:3034, in Expr.series(self, x, x0, n, dir, logx, cdir)
   3032 from sympy.series.order import Order
   3033 if n is not None:  # nseries handling
-> 3034     s1 = self._eval_nseries(x, n=n, logx=logx, cdir=cdir)
   3035     o = s1.getO() or S.Zero
   3036     if o:
   3037         # make sure the requested order is returned

File /usr/lib/python3.11/site-packages/sympy/functions/special/bessel.py:714, in besselk._eval_nseries(self, x, n, logx, cdir)
    712 newn = ceiling(n/exp)
    713 bn = besseli(nu, z)
--> 714 a = ((-1)**(nu - 1)*log(z/2)*bn)._eval_nseries(x, n, logx, cdir)
    716 b, c = [], []
    717 t = (_mexpand(r**2) + o).removeO()

File /usr/lib/python3.11/site-packages/sympy/core/mul.py:1985, in Mul._eval_nseries(self, x, n, logx, cdir)
   1983 if n0.is_nonnegative:
   1984     n0 = S.Zero
-> 1985 facs = [t.nseries(x, n=ceiling(n-n0), logx=logx, cdir=cdir) for t in self.args]
   1986 from sympy.simplify.powsimp import powsimp
   1987 res = powsimp(self.func(*facs).expand(), combine='exp', deep=True)

File /usr/lib/python3.11/site-packages/sympy/core/mul.py:1985, in <listcomp>(.0)
   1983 if n0.is_nonnegative:
   1984     n0 = S.Zero
-> 1985 facs = [t.nseries(x, n=ceiling(n-n0), logx=logx, cdir=cdir) for t in self.args]
   1986 from sympy.simplify.powsimp import powsimp
   1987 res = powsimp(self.func(*facs).expand(), combine='exp', deep=True)

File /usr/lib/python3.11/site-packages/sympy/core/expr.py:3407, in Expr.nseries(self, x, x0, n, dir, logx, cdir)
   3405     return self.series(x, x0, n, dir, cdir=cdir)
   3406 else:
-> 3407     return self._eval_nseries(x, n=n, logx=logx, cdir=cdir)

File /usr/lib/python3.11/site-packages/sympy/functions/special/bessel.py:583, in besseli._eval_nseries(self, x, n, logx, cdir)
    581 for k in range(1, (newn + 1)//2):
    582     term *= t/(k*(nu + k))
--> 583     term = (_mexpand(term) + o).removeO()
    584     s.append(term)
    585 return Add(*s) + o

File /usr/lib/python3.11/site-packages/sympy/core/decorators.py:236, in _SympifyWrapper.make_wrapped.<locals>._func(self, other)
    234 if not isinstance(other, expectedcls):
    235     return retval
--> 236 return func(self, other)

File /usr/lib/python3.11/site-packages/sympy/core/decorators.py:106, in call_highest_priority.<locals>.priority_decorator.<locals>.binary_op_wrapper(self, other)
    104         if f is not None:
    105             return f(self)
--> 106 return func(self, other)

File /usr/lib/python3.11/site-packages/sympy/core/expr.py:198, in Expr.__add__(self, other)
    195 @sympify_return([('other', 'Expr')], NotImplemented)
    196 @call_highest_priority('__radd__')
    197 def __add__(self, other):
--> 198     return Add(self, other)

File /usr/lib/python3.11/site-packages/sympy/core/cache.py:72, in __cacheit.<locals>.func_wrapper.<locals>.wrapper(*args, **kwargs)
     69 @wraps(func)
     70 def wrapper(*args, **kwargs):
     71     try:
---> 72         retval = cfunc(*args, **kwargs)
     73     except TypeError as e:
     74         if not e.args or not e.args[0].startswith('unhashable type:'):

File /usr/lib/python3.11/site-packages/sympy/core/operations.py:98, in AssocOp.__new__(cls, evaluate, _sympify, *args)
     95 if len(args) == 1:
     96     return args[0]
---> 98 c_part, nc_part, order_symbols = cls.flatten(args)
     99 is_commutative = not nc_part
    100 obj = cls._from_args(c_part + nc_part, is_commutative)

File /usr/lib/python3.11/site-packages/sympy/core/add.py:367, in Add.flatten(cls, seq)
    364 for t in newseq:
    365     for o in order_factors:
    366         # x + O(x) -> O(x)
--> 367         if o.contains(t):
    368             t = None
    369             break

File /usr/lib/python3.11/site-packages/sympy/core/cache.py:72, in __cacheit.<locals>.func_wrapper.<locals>.wrapper(*args, **kwargs)
     69 @wraps(func)
     70 def wrapper(*args, **kwargs):
     71     try:
---> 72         retval = cfunc(*args, **kwargs)
     73     except TypeError as e:
     74         if not e.args or not e.args[0].startswith('unhashable type:'):

File /usr/lib/python3.11/site-packages/sympy/series/order.py:453, in Order.contains(self, expr)
    450                 return rv
    452 obj = self.func(expr, *self.args[1:])
--> 453 return self.contains(obj)

File /usr/lib/python3.11/site-packages/sympy/core/cache.py:72, in __cacheit.<locals>.func_wrapper.<locals>.wrapper(*args, **kwargs)
     69 @wraps(func)
     70 def wrapper(*args, **kwargs):
     71     try:
---> 72         retval = cfunc(*args, **kwargs)
     73     except TypeError as e:
     74         if not e.args or not e.args[0].startswith('unhashable type:'):

File /usr/lib/python3.11/site-packages/sympy/series/order.py:428, in Order.contains(self, expr)
    426 for s in common_symbols:
    427     from sympy.series.limits import Limit
--> 428     l = Limit(ratio, s, point).doit(heuristics=False)
    429     if not isinstance(l, Limit):
    430         l = l != 0

File /usr/lib/python3.11/site-packages/sympy/series/limits.py:363, in Limit.doit(self, **hints)
    361                 return S.ComplexInfinity
    362         else:
--> 363             raise NotImplementedError("Not sure of sign of %s" % ex)
    365 # gruntz fails on factorials but works with the gamma function
    366 # If no factorial term is present, e should remain unchanged.
    367 # factorial is defined to be zero for negative inputs (which
    368 # differs from gamma) so only rewrite for positive z0.
    369 if z0.is_extended_positive:

NotImplementedError: Not sure of sign of 4 - n
```

anutosh491 commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -166,7 +168,22 @@ def test_besselk_series():
+    assert besselk(S(1)/2, x**2).series(x, n=2) == sqrt(pi/2)/x \
+        - sqrt(pi/2)*x + O(x**2)
+    assert besselk(-S(1)/2, x).series(x) == besselk(S(1)/2, x).series(x)
+    assert besselk(-S(7)/6, x).series(x) == besselk(S(7)/6, x).series(x)
```
Okay so I have gone through the changes and the tests implemented based on those changes and everything looks OK to me (hence I'll approve the PR from my side). 
I'll also ask for reviews from @jksuom who might be interested in this and has also reviewed work on series/limit methods for bessel functions in the past.

anutosh491 commented:
cc @jksuom this looks good to me. Waiting for a set of reviews from your side.

Kishore96in commented:
> cc @jksuom this looks good to me. Waiting for a set of reviews from your side.

They don't seem to have responded. How do we proceed?

Kishore96in commented:
Ping.

anutosh491 commented:
Hey @arnabnandikgp @oscarbenjamin I had reviewed & approved this PR in Sept 2023 and it looked good to me back then (there haven't been any changes after that).

Arnab could you look for errors if any. If not we can merge this !

anutosh491 commented:
Once that is done let's frame a similar PR to supercede https://github.com/sympy/sympy/pull/25840 
and fix the expansion for `bessely` too

oscarbenjamin commented on sympy/functions/special/bessel.py:
```diff
@@ -677,64 +677,79 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
-            term_three = (-1)**nu*(z/2)**nu/(2*factorial(nu))*(digamma(nu + 1) - S.EulerGamma)
-            arg = Add(*[term_one, term_two, term_three]).as_leading_term(x, logx=logx)
-            return arg
+            if nu is S.Zero:
+                # Equation 9.6.8 of Abramowitz and Stegun (10th ed, 1972).
+                term = -log(z) - S.EulerGamma + log(2)
+            else:
+                # Equation 9.6.9 of Abramowitz and Stegun (10th ed, 1972).
+                term = gamma(Abs(nu))*(z/2)**(-Abs(nu))/2
```
This is checking if `nu is S.Zero` but what if `nu` is an expression that is possibly zero?

Is it possible to handle that case or do we need to know definitely whether `nu` is zero or not?

anutosh491 replied:
I think we need to know before,  if `nu` is `zero` before proceeding or not here, so I think this works.
And also not sure if we were looking for catering those cases where `nu` might be an expression here.

So basically my PR (https://github.com/sympy/sympy/pull/23844) implemented basic series for bessel functions for example `bessely(1, sin(x))` but I forgot to consider the important case that `nu` might be non-integer which this PR is fixing.

oscarbenjamin replied:
So what happens if `nu` is just a symbol?

Does it just return a series that will be valid assuming that the symbol is not zero?

Neglecting a degenerate case like the symbol being zero does not seem completely unreasonable to me but we need to clear if that is the expected behaviour or not.

Lower down there is:
```
if nu.is_integer:
   # integer case
else:
   # non-integer case
```
This is more problematic. For a symbol `nu.is_integer` will likely return None and this will take the non-integer case.

Do we want to use the non-integer case whenever `nu` is a symbol?

You might notice around the codebase things like this to handle three-valued logic:
```python
if nu.is_integer:
   # integer case
elif nu.is_integer is False:
   # non-integer case
else:
   # Here we do not know if it is an integer or not.
   # We need to decide how to handle this somehow.
   # Ideally this case should not return likely incorrect results.
   # Perhaps we could raise an error or perhaps there is
   # some other routine that we can fall back on or perhaps
   # the result could be unevaluated or something.
```

anutosh491 replied:
Ahh yes, we should be tackling this. Just wasn't sure if we should do it through this pr or not.

Yeah I think we return a general series for besselk(nu, m) not involving any checks for zero etc

Kishore96in replied:
I have now modified the `nu is S.Zero` block to raise a `NotImplementedError` if we do not know whether the symbol is zero or nonzero. Similarly, I have done
```
if nu.is_integer:
   # integer case
elif nu.is_noninteger:
   # non-integer case
#fallback appears later on
```

Series expansion currently (even on master) raises a `NotImplementedError` error if the order of the bessel function is symbolic (see https://github.com/sympy/sympy/pull/25664#issuecomment-1732226476).

I am not sure about how to correctly implement series expansion for symbolic order, given that for Bessel functions, the number of terms in the series depends on the order (so it does not seem it can be handled by simply returning a `Piecewise` instance for the integer and non-integer cases).

oscarbenjamin replied:
> I am not sure about how to correctly implement series expansion for symbolic order, given that for Bessel functions, the number of terms in the series depends on the order (so it does not seem it can be handled by simply returning a `Piecewise` instance for the integer and non-integer cases).

Okay well let's make sure at least that we error in the correct way rather than returning incorrect results then.

Although can it not be a Piecewise series like:
```python
In [1]: Piecewise((exp(x).series(n=3), x>0), (sin(x).series(n=3), True))
Out[1]: 
⎧         2                   
⎪        x     ⎛ 3⎞           
⎪1 + x + ── + O⎝x ⎠  for x > 0
⎨        2                    
⎪                             
⎪         ⎛ 3⎞                
⎩    x + O⎝x ⎠       otherwise
```

Kishore96in replied:
I am now trying to modify `besselk._eval_nseries` to return a Piecewise, which I believe forces me to implement series expansion for all possible cases (I can't raise `NotImplementedError`). I have a question about the expected output for series like
![equation 9.7.2 of Abramowitz and Stegun](https://github.com/user-attachments/assets/61f5cadb-58e1-4259-bd73-8f06c0f294e5)
(this is for expansion about z=infinity, which would correspond to the currently unimplemented `exp < 0` branch of `besselk._eval_nseries`). Is it okay to return unexpanded prefactors like the `exp(-z)` in the equation above? I think this is what users would expect, but are there any assumptions made elsewhere in the code (e.g. that the output is a polynomial in x) that this would break?

Note that
```
In [100]: ( exp(-x) ).series(x, oo, n=4)
Out[100]: 
 ⎛1        ⎞
O⎜──; x → ∞⎟
 ⎜ 4       ⎟
 ⎝x        ⎠
```

oscarbenjamin replied:
It might not be a good idea to use a Piecewise here. As @anutosh491 said above:

> Ahh yes, we should be tackling this. Just wasn't sure if we should do it through this pr or not.

I'm just musing about the possibilities. For now let's just make sure that the conditions are being checked properly before returning the series in each case.

anutosh491 replied:
yes @Kishore96in, we might not want to address everything through this PR. Just have branches for all cases and we can return "NotImplemented" there. I do like most of the work here and I did approve the standard non-integer cases a while back too !

Kishore96in replied:
> > I am not sure about how to correctly implement series expansion for symbolic order, given that for Bessel functions, the number of terms in the series depends on the order (so it does not seem it can be handled by simply returning a `Piecewise` instance for the integer and non-integer cases).
> 
> Okay well let's make sure at least that we error in the correct way rather than returning incorrect results then.
> 

With my latest changes,
```
In [8]: besselk(y,x).series(x)
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[8], line 1
----> 1 besselk(y,x).series(x)

File /usr/lib/python3.12/site-packages/sympy/core/expr.py:3020, in Expr.series(self, x, x0, n, dir, logx, cdir)
   3017 if x.is_positive is x.is_negative is None or x.is_Symbol is not True:
   3018     # replace x with an x that has a positive assumption
   3019     xpos = Dummy('x', positive=True)
-> 3020     rv = self.subs(x, xpos).series(xpos, x0, n, dir, logx=logx, cdir=cdir)
   3021     if n is None:
   3022         return (s.subs(xpos, x) for s in rv)

File /usr/lib/python3.12/site-packages/sympy/core/expr.py:3028, in Expr.series(self, x, x0, n, dir, logx, cdir)
   3026 from sympy.series.order import Order
   3027 if n is not None:  # nseries handling
-> 3028     s1 = self._eval_nseries(x, n=n, logx=logx, cdir=cdir)
   3029     o = s1.getO() or S.Zero
   3030     if o:
   3031         # make sure the requested order is returned

File /usr/lib/python3.12/site-packages/sympy/functions/special/bessel.py:785, in besselk._eval_nseries(self, x, n, logx, cdir)
    783         return Add(*a) + Add(*b) + o
    784     else:
--> 785         raise NotImplementedError("besselk expansion is only implemented for real order")
    787 return super(besselk, self)._eval_nseries(x, n, logx, cdir)

NotImplementedError: besselk expansion is only implemented for real order
```
Is the error message okay?

oscarbenjamin commented on sympy/functions/special/bessel.py:
```diff
@@ -677,64 +677,82 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
+                    term = p*(digamma(k + nu + 1) + digamma(k + 1))
+                    c.append(term)
+                return a + Add(*b) + Add(*c) # Order term comes from a
+            elif nu.is_noninteger:
```
This is not the same as `nu.is_integer is False` because `noninteger` implies `real`. Would it be acceptable here for `nu` to be a complex number like `1+I` or is it important that we only allow real `nu` here?

Kishore96in replied:
While the formula in the given reference is correct (I think) for complex numbers, my implementation assumes nu is real. I have now added a comment mentioning this, and explicitly raised NotImplementedError for non-real order.

oscarbenjamin replied:
How exactly does the implementation assume that `nu` is real?

Is it just the `ceiling((n+nu)/exp)` part?

Is it not easy to fix that so that it is valid for complex `nu`?

Ideally if there is a general expression that gives the series with all analytic coefficients then we would use it because it makes handling all of these different cases a lot simpler.

Kishore96in replied:
![Wolfram general expression](https://github.com/user-attachments/assets/facf4421-8750-4a56-95e4-12dc87790ddd)

The above is the general expression from the Wolfram Mathworld link (implemented in the `nu.is_noninteger` branch). Note the $z^{\nu}$ and $z^{-\nu}$ in the two terms. For a power series expansion, we want the powers of $z$ to be real, and so directly putting complex $\nu$ into the series above is not useful. Nevertheless, the expression given above is still formally correct for complex order (just not useful as a power series expansion).

The trick used by Dunster (1990) is that when $\nu$ is purely imaginary (say $\nu = i \mu$, with $\mu$ being real), you can write $z^{i\mu} - z^{-i\mu} = 2 i \sin(\mu\log z)$, giving, after some manipulation, the series
![dunster 1990 eq 2.8](https://github.com/user-attachments/assets/e2c1296f-72bb-4507-aebe-93beca491198)

Thus, while the series I have used for the implementation of non-integer order is mathematically valid for non-real order as well, I don't see how I could implement the above handling of, e.g., purely imaginary order in `_eval_nseries` without making a new conditional branch.

Kishore96in replied:
I've updated the comment 'for general complex order' -> 'for non-real order as well'

Kishore96in committed:
besselk: add tests for series expansion, non-integer order

Kishore96in commented:
Rebased and resolved a merge conflict.

anutosh491 commented:
I like most of the work here and this might be close to what we want. @arnabnandikgp could you give this a thorough review (verifying the cases and also confirming the tests) because I would like you to address the same issue for **bessely** (https://github.com/sympy/sympy/pull/25840) and that would require the exact same structure.

cc @oscarbenjamin would be great if you could give this a look too.

oscarbenjamin commented on sympy/functions/special/bessel.py:
```diff
@@ -698,64 +698,91 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
+                # of purely imaginary order, with an application to second-order
+                # linear differential equations having a large parameter".
+                # SIAM J. Math. Anal. Vol 21, No. 4, pp 995-1018.], the
+                # implementation here assumes nu is real.
```
Perhaps rather than saying "the implementation here assumes nu is real" this could say "if nu is not real then the expression would not represent a power series because it would have powers of x with non-real exponent."

Kishore96in replied:
Done. New wording is "While the expression in the reference above seems correct for non-real order as well, it would need some manipulation (not implemented) to be written as a power series in x with real exponents".

oscarbenjamin commented:
I haven't checked any of the results here in detail but otherwise it looks okay to me.

arnabnandikgp commented:
Just went through the tests, seems good to me.

anutosh491 commented:
> I haven't checked any of the results here in detail but otherwise it looks okay to me.

Have already approved it and confirmed the results :)
Go ahead and merge once you think its good.

oscarbenjamin commented:
> Go ahead and merge once you think its good.

My statement

> I haven't checked any of the results here in detail but otherwise it looks okay to me.

is perhaps not very explicit so let me rephrase:

I checked a couple of things related to previous discussion but I have not otherwise checked it carefully and I don't intend to because I don't want to take responsibility for checking all PRs in detail. If you are happy and have checked it in detail then go ahead and merge.

anutosh491 commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -172,6 +180,8 @@ def test_besselk_series():
         S.EulerGamma/2) + x**(S(5)/2)*(log(x)/24 - S(5)/36 + S.EulerGamma/12) + O(x**3*log(x))
     assert besselk(-2, sin(x)).series(x, n=4) == besselk(2, sin(x)).series(x, n=4)
+    assert besselk(2, x**2).series(x, n=2) == 2/x**4 + O(x**2)
+    assert (x**2*besselk(2, x)).series(x, n=2) == 2 + O(x**2)
```
Can we check for more terms here ?

Kishore96in replied:
Added another test which goes up to n=6. Turns out the RHS of one of the tests above was wrong, so I have fixed that and retained this test.

anutosh491 commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -180,6 +190,20 @@ def test_besselk_series():
+        sqrt(pi/2)*x**(S(1)/4) + sqrt(pi/2)*x**(S(3)/4)/2 - \
+            sqrt(pi/2)*x**(S(5)/4)/6 + sqrt(pi/2)*x**(S(7)/4)/24 + O(x**2)
+    assert besselk(S(1)/2, x**2).series(x, n=2) == sqrt(pi/2)/x \
+        - sqrt(pi/2)*x + O(x**2)
```
n=2 seems fine here as we are looking at fractional powers.

anutosh491 commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -69,6 +69,14 @@ def test_besselk_leading_term():
     assert besselk(0, x).as_leading_term(x) == -log(x) - S.EulerGamma + log(2)
```
We need to test all cases extensively specially for the leading term method, otherwise it'll be pretty easy to mess up simple limits.
For example this is one I could find out from just trying 
```
>>> besselk(0, cos(x)).as_leading_term(x)
besselk(0, cos(x))    # should be besselk(0, 1)
>>> besselk(0, cos(x)).series(x, n = 4)
besselk(0, 1) + x**2*besselk(1, 1)/2 + O(x**4)
```
The series looks perfectly fine though

anutosh491 replied:
cc @arnabnandikgp I would really you to help @Kishore96in with the tests here. No harm in doing in extensive testing because as far as I know we've had poor output for limits/series for bessel functions (especially the modified ones)

It would help with your PR too.

anutosh491 replied:
This is the `e.is_zero` cases (after going through the positive and negative) case

Kishore96in replied:
@anutosh491 I have now fixed the `e.is_zero` case. Also added tests for this and the e<0 case.

```
In [1]: besselk(1,cos(x)).as_leading_term(x)
Out[1]: besselk(1, 1)

In [2]: besselk(3,1/x).as_leading_term(x)
Out[2]: 
       -1 
       ───
        x 
√2⋅√π⋅ℯ   
──────────
      ___ 
     ╱ 1  
2⋅  ╱  ─  
  ╲╱   x  

In [9]: besselk(3,1/sin(x)).as_leading_term(x)
Out[9]: 
       -1 
       ───
        x 
√2⋅√π⋅ℯ   
──────────
      ___ 
     ╱ 1  
2⋅  ╱  ─  
  ╲╱   x  


```
Unsure about the last one, since it disagrees with @arnabnandikgp 's implementation of `besselk._eval_aseries`
```In [8]: besselk(3,1/sin(x)).series(x, n=1).expand().removeO()
Out[8]: 
        -1    
       ────── 
       sin(x) 
√2⋅√π⋅ℯ       
──────────────
      ________
     ╱   1    
2⋅  ╱  ────── 
  ╲╱   sin(x)
```

anutosh491 replied:
Yess this is an error. Let's raise an issue for the same

arnabnandikgp replied:
I am satisfied with the tests added in the recent commit.

anutosh491 commented:
Need more extensive testing for the series too. I checked some cases, mostly things look fine.
Not too sure about this though 
```
>>> besselk(1, 1/cos(x)).series(x, n = 3)
besselk(1, 1) + x**2*(-besselk(2, 1)/4 - besselk(0, 1)/4) + O(x**3)
```
This is what wolfram had to say 
```
BesselK[1, 1] + (x^2 (-BesselK[1, 1] + BesselK[2, 1]))/2 + O[x]^3
```
Don't think these are the same. Are they ?

Kishore96in commented:
> Need more extensive testing for the series too. I checked some cases, mostly things look fine. Not too sure about this though
> 
> ```
> >>> besselk(1, 1/cos(x)).series(x, n = 3)
> besselk(1, 1) + x**2*(-besselk(2, 1)/4 - besselk(0, 1)/4) + O(x**3)
> ```
> 
> This is what wolfram had to say
> 
> ```
> BesselK[1, 1] + (x^2 (-BesselK[1, 1] + BesselK[2, 1]))/2 + O[x]^3
> ```
> 
> Don't think these are the same. Are they ?

@anutosh491  I think Wolfram is wrong here, which one can check by plotting the series (isympy session):
```
In [15]: expr = besselk(1,1/cos(x)).series(x, n=3).removeO()

In [18]: exprW = besselk(1,1) + x**2*(-besselk(1,1) + besselk(2,1))/2

In [30]: plot(besselk(1,1/cos(x)), exprW, expr, (x,-1,1), label=("actual", "
    ...: wolfram", "ours"))
Out[30]: <sympy.plotting.backends.matplotlibbackend.matplotlib.MatplotlibBackend at 0x7d54b6fb89b0>

In [31]: plt.legend()
Out[31]: <matplotlib.legend.Legend at 0x7d54b6142750>
```

![wolfram_wrong](https://github.com/user-attachments/assets/826b8315-60b4-413d-8f56-6cc39bf0b75a)

oscarbenjamin commented:
> Well that's a first :)

No, it is not. It is useful to compare against Wolfram and other computer algebra systems but none of them should be trusted automatically. For one thing they may define functions and operations differently. Likewise SymPy's plot function should not be trusted automatically especially since it uses a number of other libraries like SciPy etc. The gold standard in SymPy is typically `.evalf()` which usually shows more or less directly what mpmath does. That is not always correct but mostly amounts to defining which each function is in SymPy.

arnabnandikgp commented:
Ohk, i just went through the expressions given by both sympy and mathematica and in mathematica we have
![image](https://github.com/user-attachments/assets/186e4861-2c0f-463c-93a9-98f592b1a63d)
i.e 
```
exprW = besselk(1,1) + x**2*(besselk(1,1) - besselk(2,1))/2 + O(x**3)
``` 
(anutosh mistakenly flipped the sign of the coefficient of x**2)
 and from sympy we have 
```python
In [15]: besselk(1, 1/cos(x)).series(x, n=3)
Out[15]: 
                 2 ⎛  besselk(2, 1)   besselk(0, 1)⎞    ⎛ 3⎞
besselk(1, 1) + x ⋅⎜- ───────────── - ─────────────⎟ + O⎝x ⎠
                   ⎝        4               4      ⎠        

In [16]: _.n()
Out[16]: 
                                       2    ⎛ 3⎞
0.601907230197235 - 0.511465834218971⋅x  + O⎝x ⎠

```
and if we compare it with the value of given by mathematica i.e
```python
In [19]: exprW = besselk(1,1) + x**2*(besselk(1,1) - besselk(2,1))/2

In [20]: exprW.n()
Out[20]: 
                                       2
0.601907230197235 - 0.511465834218971⋅x 

```
same is confirmed by the plot
![image](https://github.com/user-attachments/assets/ca7f599a-86d3-4988-94ea-2b93c4d58331)

arnabnandikgp commented:
@Kishore96in just rebase your branch onto the current master to fix the code quality test.
@anutosh491 let's get this in, this has been held back for so long.

Kishore96in commented on sympy/functions/special/bessel.py:
```diff
@@ -698,64 +698,92 @@ def _eval_as_leading_term(self, x, logx, cdir):
+                    b.append(term)
+                    for k in range(1, nu):
+                        denom = (k - nu)*k
+                        if denom == S.Zero:
```
@anutosh491 This condition (`denom == S.Zero`) was inherited from your implementation for non-integer order, but it seems to me that it should never be true (by adding an assert, one can at least see that it is not triggered by any of the tests). Can you think of any cases in which it is possible to have `denom == S.Zero`?

anutosh491 replied:
Hmmm, well if there's a case ... We should surely have a test for it. Will have to look if this would be necessary.

Arnab, can you think of something here. I'll think about it in a while too !

Kishore96in replied:
I've removed that branch for now (and so we assume `denom != 0`); will add it back if we find an example.

oscarbenjamin commented:
> and if we compare it with the value of given by mathematica i.e
> 
> ```python
> In [19]: exprW = besselk(1,1) + x**2*(besselk(1,1) - besselk(2,1))/2
> 
> In [20]: exprW.n()
> Out[20]: 
>                                        2
> 0.601907230197235 - 0.511465834218971⋅x 
> ```

The main thing to compare is the numerical evaluation of the original function whose series is being computed i.e.:
```python
In [2]: besselk(1,1/cos(x)).series(x, n=3).removeO().evalf(subs={x:S(1)/10})
Out[2]: 0.596792571855045

In [3]: besselk(1,1/cos(x)).evalf(subs={x:S(1)/10})
Out[3]: 0.596799105027444
```
We want to compare the series result with the non-series function whose series was computed.

Kishore96in commented:
BTW, there seems to be some issue with the `O` object. Naively, I would expect the two commands that follow to give the same output, but that is not the case.
```
In [4]: O(x) + O(x*log(x))
Out[4]: O(x⋅log(x))

In [5]: O(x*log(x) + x)
Out[5]: O(x⋅(log(x) + 1))
```
I only noticed this because I was trying to get the right form of the `O(...)` terms on the RHS of the tests.

oscarbenjamin commented:
> BTW, there seems to be some issue with the `O` object.

There are many issues with the `O` object:
```python
In [1]: 0 * O(0)
Out[1]: O(0)
```

anutosh491 commented:
> Well that's a first :)

> No, it is not

Arggg, correcting myself. I wasn't speaking for whole of SymPy just the series module. During my GSoC and afterwards, Kalevi and I always used to check for results against wolfram and I don't think I've witnessed a case where a series expansion or limit was solved better through SymPy than Wolfram.

anutosh491 commented:
Ok I am satisfied with what we have here. Some comments and roadmap to proceed (cc @arnabnandikgp @Kishore96in )

1) I think there are some issues being pointed out about the Order term. Arnab could you raise issues for the same. Let's keep track of what doesn't work
2) I remember `denom == S.Zero` having some use case. Forgot what it was about or if we even tested it out.
3) Arnab needs to address https://github.com/sympy/sympy/issues/27036
4) Then we can fix the same for `bessely`

@Kishore96in thanks a lot for working on this. Would be great if you could squash the 12 commits into a couple meaningful ones.

anutosh491 commented:
It's fine, just wanted to make note of it if we end up with a problem in the future. Please go ahead and squash the commits into few meaningful ones and then I can merge this :)

Kishore96in commented:
@anutosh491 Squashed and rebased.

oscarbenjamin commented:
Good work. Thanks everyone!
